lib/identity_cache/query_api.rb

```diff
@@ -229,6 +230,7 @@ def require_if_necessary #:nodoc:
       def resolve_cache_miss(id)
         record = self.includes(cache_fetch_includes).reorder(nil).where(primary_key => id).first
         preload_id_embedded_associations([record]) if record
+        record.try(:readonly!) if IdentityCache.fetch_read_only_records
```

Comment:
I dislike `try` because it is poorly documented and has changed behaviour since early versions of rails.  It effectively does `record.respond_to?(:readonly!) ? record : nil` which subtly stops working if the `readonly!` method is removed or renamed.  Usually `try!` is more appropriate, which effectively does `record.nil? ? nil : nil`

In this case I would actually wrap this line and the above line in a `if record` condition.

I think we want to also avoid marking this record as readonly in a transaction

```ruby
    if record
        preload_id_embedded_associations([record])
        record.readonly! if IdentityCache.fetch_read_only_records && IdentityCache.should_use_cache?
    end
```